### PR TITLE
[MRG] Serializing models

### PR DESCRIPTION
**Pull request by kushalkolar**
Cleaner followup from PR #160

To do:

- [x] Add to PAA and SAX models
- [x] Use synthetic data for tests
- [x] Add to docs gallery
- [x] Add to CHANGELOG

### CODE REVIEW

**Comment by kushalkolar** [comment]
tests are passing except for the ones without JIT https://travis-ci.org/rtavenar/tslearn/builds/640113360

I'll do the small remaining things tomorrow.

**Comment by rtavenar** [comment]
The ones that dont pass just time out, so using small synthetic datasets should fix this.

File: tslearn/tests/test_serialize_models.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    # Keep first 3 classes
+    X_train = X_train[y_train < 4]
+    numpy.random.shuffle(X_train)
+    # Keep only 50 time series
+    X_train = TimeSeriesScalerMeanVariance().fit_transform(X_train[:50])
+    sz = X_train.shape[1]
+
+    gak_km = GlobalAlignmentKernelKMeans(n_clusters=3,
+                                         sigma=sigma_gak(X_train),
+                                         n_init=20,
```
Comment (johannfaouzi) [inline tslearn/tests/test_serialize_models.py]: `n_init=1` is enough, we don't want to check that the results are good, we just want to check that there are consistent

File: tslearn/tests/test_serialize_models.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    n_queries = 2
+    n_neighbors = 4
+
+    knn = KNeighborsTimeSeries(n_neighbors=n_neighbors)
+
+    _check_not_fitted(knn)
+
+    knn.fit(X_train)
+
+    # serialize to all all_formats
```
Comment (johannfaouzi) [inline tslearn/tests/test_serialize_models.py]: Is it possible to use `_check_params_predict()` here or not?

Comment (kushalkolar) [reply tslearn/tests/test_serialize_models.py]: thanks for the suggestion, I generalized `_check_params_predict()` now.

File: tslearn/piecewise.py [hunk 3]
```diff
@@ -199,9 +218,9 @@ def distance_paa(self, paa1, paa2):
         .. [1] E. Keogh & M. Pazzani. Scaling up dynamic time warping for
            datamining applications. SIGKDD 2000, pp. 285--289.
         """
-        if self.size_fitted_ < 0:
-            raise ValueError("Model not fitted yet: cannot be used for " +
-                             "distance computation.")
+        if not self._is_fitted():
```
Comment (kushalkolar) [inline tslearn/piecewise.py]: check with `_is_fitted()`, corresponding tests changed as well.

File: tslearn/piecewise.py [hunk 4]
```diff
@@ -311,8 +331,34 @@ class SymbolicAggregateApproximation(PiecewiseAggregateApproximation):
     def __init__(self, n_segments, alphabet_size_avg):
         PiecewiseAggregateApproximation.__init__(self, n_segments)
         self.alphabet_size_avg = alphabet_size_avg
-        self.breakpoints_avg_ = _breakpoints(self.alphabet_size_avg)
```
Comment (kushalkolar) [inline tslearn/piecewise.py]: Moved these to `fit()`

File: tslearn/piecewise.py [hunk 5]
```diff
@@ -570,7 +638,7 @@ def fit_transform(self, X, y=None, **fit_params):
             (standard SAX symbols) and the last d are for slopes
         """
         X_ = to_time_series_dataset(X)
-        return self._fit(X_)._transform(X_)
+        return self.fit(X_)._transform(X_)
```
Comment (kushalkolar) [inline tslearn/piecewise.py]: since `fit()` now sets some of the model params

File: tslearn/piecewise.py [hunk 6]
```diff
@@ -199,9 +218,9 @@ def distance_paa(self, paa1, paa2):
         .. [1] E. Keogh & M. Pazzani. Scaling up dynamic time warping for
            datamining applications. SIGKDD 2000, pp. 285--289.
         """
-        if self.size_fitted_ < 0:
-            raise ValueError("Model not fitted yet: cannot be used for " +
-                             "distance computation.")
+        if not self._is_fitted():
+            raise NotFittedError("Model not fitted yet: cannot be used for " +
```
Comment (kushalkolar) [inline tslearn/piecewise.py]: raise `NotFittedError` instead

**Comment by kushalkolar** [comment]
Everything is pretty much done.

I moved model parameter declarations in PAA and SAX to the `fit()` method. Their `fit_transform()` method now calls `fit()` instead of `_fit()` so that these model declarations are also made over there. 

waiting to see if tests pass on travis https://travis-ci.org/rtavenar/tslearn/builds/640745373

**Comment by kushalkolar** [comment]
yay tests are passing, @rtavenar what do you think about making `h5py` an optional dependency? raise warning when `hdftools` is imported if `h5py` is not installed, and exception when user tries to use `to_hdf5` or `from_hdf5`

**Comment by kushalkolar** [comment]
Added notes about `h5py` to the `README` and the `to_hdf5`/`from_hdf5` docstrings. Latest tests are passing https://travis-ci.org/rtavenar/tslearn/builds/641069143

**Comment by rtavenar** [comment]
Thanks @kushalkolar for this great contribution!

Now merging to `dev`, and we are ready for tslearn 0.3!

**Pull request merged**